src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,319 @@
+        private const int BlockSize = 16 * 1024;
+
+        internal MultiMemory(byte[][] blocks, int start, int length)
+        {
+            if (length == 0)
+            {
+                _blocks = Array.Empty<byte[]>();
+                _start = 0;
+                _length = 0;
+            }
+            else
+            {
+                Debug.Assert(start >= 0);
+                Debug.Assert(length >= 0);
+                Debug.Assert(start + length <= blocks.Length * BlockSize);
+
+                _blocks = blocks;
+                _start = start;
+                _length = length;
+            }
+        }
+
+        private static int GetBlockIndex(int offset) => offset / BlockSize;
+        private static int GetOffsetInBlock(int offset) => offset % BlockSize;
+
+        public int Length => _length;
+
+        public ref byte this[int index]
+        {
+            get
+            {
+                if (index < 0 || index >= _length)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+
+                int offset = _start + index;
+                return ref _blocks[GetBlockIndex(offset)][GetOffsetInBlock(offset)];
+            }
+        }
+
+        public int BlockCount => _length == 0 ? 0 : GetBlockIndex(_start + _length - 1) - GetBlockIndex(_start) + 1;
+
+        public Memory<byte> GetBlock(int blockIndex)
+        {
+            if (blockIndex < 0 || blockIndex >= BlockCount)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            Debug.Assert(_length > 0, "Length should never be 0 here because BlockCount would be 0");
+
+            int startInBlock = (blockIndex == 0 ? GetOffsetInBlock(_start) : 0);
+            int endInBlock = (blockIndex == BlockCount - 1 ? GetOffsetInBlock(_start + _length - 1) + 1 : BlockSize);
+
+            Debug.Assert(0 <= startInBlock, $"Invalid startInBlock={startInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+            Debug.Assert(startInBlock < endInBlock, $"Invalid startInBlock={startInBlock}, endInBlock={endInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+            Debug.Assert(endInBlock <= BlockSize, $"Invalid endInBlock={endInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+
+            return new Memory<byte>(_blocks[GetBlockIndex(_start) + blockIndex], startInBlock, endInBlock - startInBlock);
+        }
+
+        public MultiMemory Slice(int start)
+        {
+            if (start < 0 || start > _length)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            return new MultiMemory(_blocks, _start + start, _length - start);
+        }
+
+        public MultiMemory Slice(int start, int length)
+        {
+            if (start < 0 || length < 0 || start + length > _length)
```

Comment:
`start + length` can overflow.

```suggestion
            if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
```

Or tho eke out more perf, one could (on 64-bit) bring it into `ulong`-domain and compare over there. But I don't think it's worth it.

<details>
  <summary>code</summary>

```c#
if (Environment.Is64BitProcess)
{
    if ((ulong)(uint)start + (ulong)(uint)length > (ulong)(uint)_length)
    {
        throw new IndexOutOfRangeException();
    }
}
else
{
    if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
    {
        throw new IndexOutOfRangeException();
    }
}
```
</details>